### PR TITLE
[Task] Add and enroll habits feature has been added

### PR DESCRIPTION
**Pull request by luchward**
**Restrictions:**

- refactoring of calendar components is out of scope;
- writing tests is out of scope.

**Results:**

- user story [#1000](https://github.com/ita-social-projects/GreenCity/issues/1000) has been completed;
- components that display the profile page have been refactored to meet mockup requirements;
- profile-dashboard component displays habits in progress and acquired;
- profile-dashboard component lets user to enroll and unenroll his/her/its habits;
- hyperlinks to user's profiles on social networks are displayed on the profile page;
- habits are displayed in corresponding localization;
- habit-service and habit-assign-service have been created; thay implement all methods of corresponding backend controllers.

image

### CODE REVIEW

**Review by JimHolman1199 — CHANGES_REQUESTED** [review]
incorrect alignment of the whole page.
how it looks : 
Screenshot 2020-12-30 at 23 44 56
how it should be : 
Screenshot 2020-12-30 at 23 49 34

use 'container' class! 

missing space
Screenshot 2020-12-30 at 23 50 57

File: src/app/component/user/components/profile/profile-dashboard/profile-dashboard.component.ts [hunk 1]
```diff
@@ -1,50 +1,68 @@
...
+      .subscribe(() => this.executeRequests());
+  }
+
+  private executeRequests() {
+    this.loading = true;
+    this.habitAssignService.getAssignedHabits()
+      .pipe(take(1))
+      .subscribe((response: Array<HabitAssignInterface>) => {
+        const sortedHabits = response
+          .sort((a, b) => (a.habit.id > b.habit.id) ? 1 : (b.habit.id > a.habit.id) ? -1 : 0);
```
Comment (JimHolman1199) [inline src/app/component/user/components/profile/profile-dashboard/profile-dashboard.component.ts:53]: Move this array "sort" operation to a separate statement.

File: src/app/service/habit/habit.service.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  getHabitShoppingList(id: number): Observable<Array<HabitShoppingListInterface>> {
+    return this.http.get<Array<HabitShoppingListInterface>>(`${habitLink}/${id}/shopping-list?lang=${this.language}`);
+  }
+
+  getHabitsTags(): Observable<Array<string>> {
+    return this.http.get<Array<string>>(`${habitLink}/tags?lang=${this.language}`);
+  }
+
+  getHabitsByTagAndLang(page: number, size: number, sort: string = 'asc', tags: Array<string>): Observable<HabitListInterface> {
```
Comment (JimHolman1199) [inline src/app/service/habit/habit.service.ts:43]: Default parameters should be last.

**Review by JimHolman1199 — COMMENTED** [review]
consider writing unit tests because the coverage of the new code is very low

**Comment by luchward** [comment]
JimHolman1199 Thank you for your attentiveness.

However there are not mockups drawn for 1200 px size:

image

In this case I have implemented a sequence of container sizes:

- more or equal 320 px - fluid;
- more or equal 576 px - fluid;
- more or equal 768 px - 740 px;
- more or equal 1024 px - 960 px;
- more or equal 1440 px - 1140 px.

Using of container class leads to incorrect alignment:

Безіменний

and I don't wanna handle it, so I implemented my own way.

File: src/app/component/user/components/profile/profile-dashboard/profile-dashboard.component.ts (hunk 1 above)
Comment (luchward) [reply src/app/component/user/components/profile/profile-dashboard/profile-dashboard.component.ts]: fixed

File: src/app/service/habit/habit.service.ts (hunk 2 above)
Comment (luchward) [reply src/app/service/habit/habit.service.ts]: fixed

**Comment by JimHolman1199** [comment]
Screenshot 2021-01-03 at 00 23 04
Screenshot 2021-01-03 at 00 24 38
calendar is not aligned to other content

You will need to use a predefined grid system, as your page will not match other pages.
Screenshot 2021-01-03 at 00 26 24

**Review by JimHolman1199 — CHANGES_REQUESTED** [review]
fix page layout

File: src/app/component/user/components/habit/all-habits/all-habits.component.ts [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
+import { BehaviorSubject, Subscription} from 'rxjs';
 
 import { LocalStorageService } from '@global-service/localstorage/local-storage.service';
-import { ServerHabitItemPageModel } from '@global-user/models/habit-item.model';
-import { AllHabitsService } from './services/all-habits.service';
+import { HabitService } from '@global-service/habit/habit.service';
+import { map } from 'rxjs/operators';
+import { HabitInterface, HabitListInterface } from '../../../../../interface/habit/habit.interface';
+import { singleNewsImages } from '../../../../../image-pathes/single-news-images';
+import {HabitAssignService} from '@global-service/habit-assign/habit-assign.service';
```
Comment (JimHolman1199) [inline src/app/component/user/components/habit/all-habits/all-habits.component.ts:10]: Remove this unused import of 'HabitAssignService'

**Review by JimHolman1199 — CHANGES_REQUESTED** [review]

**Comment by luchward** [comment]
I know about calendars, but I can not refactor every component I touch. It is infinite task.
What concerns to layout, as I understand some mockups don't meet styleguide, but it is not my resposibility. I follow mockup requirements for the particular page.

File: src/app/component/user/components/habit/all-habits/all-habits.component.ts (hunk 3 above)
Comment (luchward) [reply src/app/component/user/components/habit/all-habits/all-habits.component.ts]: fixed

**Comment by luchward** [comment]
I came up with a solution to change the last breakpoint from 1440 to 1200 px

**Review by idzhersh — APPROVED** [review]

**Review by JimHolman1199 — APPROVED** [review]

**Review by KirillAbramo — APPROVED** [review]

**Review by ZooMingGG — APPROVED** [review]

**Review by KirillAbramo — APPROVED** [review]

**Review by idzhersh — APPROVED** [review]

**Review by JimHolman1199 — CHANGES_REQUESTED** [review]
something went wrong
Screenshot 2021-01-05 at 22 26 37

File: src/app/component/user/components/profile/profile-widget/profile-header/profile-header.component.scss [hunk 4]
```diff
@@ -1,153 +1,252 @@
...
+      color: $grey-light-color;
+      justify-self: center;
+    }
+
+    .online {
+      display: none;
+    }
+
+    .rate {
+      width: 73px;
```
Comment (mr-pavlenko) [inline src/app/component/user/components/profile/profile-widget/profile-header/profile-header.component.scss:66]: Why is the width fixed?

File: src/app/component/user/components/profile/profile-widget/profile-header/profile-header.component.scss [hunk 5]
```diff
@@ -1,153 +1,252 @@
...
+      font-size: 11px;
+      line-height: 16px;
+      color: $main-color;
+      width: 288px;
+      justify-self: center;
+    }
+
+    .social {
+      display: flex;
+      width: 288px;
```
Comment (mr-pavlenko) [inline src/app/component/user/components/profile/profile-widget/profile-header/profile-header.component.scss:93]: Why is the width fixed?

File: src/app/service/habit/habit.service.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    return this.http.get<Array<string>>(`${habitLink}/tags?lang=${this.language}`);
+  }
+
+  getHabitsByTagAndLang(page: number, size: number, tags: Array<string>, sort: string = 'asc'): Observable<HabitListInterface> {
+    return this.http.get<HabitListInterface>(`${habitLink}/tags/search?lang=${this.language}
+                                              &page=${page}&size=${size}&sort=${sort}&tags=${tags}`);
+  }
+
+  ngOnDestroy(): void {
+    this.destroy.next(null);
```
Comment (mr-pavlenko) [inline src/app/service/habit/habit.service.ts:49]: Why did you use null?

File: src/app/component/user/components/profile/profile-widget/profile-header/profile-header.component.scss (hunk 4 above)
Comment (luchward) [reply src/app/component/user/components/profile/profile-widget/profile-header/profile-header.component.scss:66]: This block works as background for the text and has fixed width due to mockup requirements

File: src/app/component/user/components/profile/profile-widget/profile-header/profile-header.component.scss (hunk 5 above)
Comment (luchward) [reply src/app/component/user/components/profile/profile-widget/profile-header/profile-header.component.scss:93]: Outer block has fixed width according to mockup and particular alignment of content, so this block has fixed width as well.

File: src/app/service/habit/habit.service.ts (hunk 6 above)
Comment (luchward) [reply src/app/service/habit/habit.service.ts:49]: my fault

**Review by yuriiLisunov — APPROVED** [review]

**Review by JimHolman1199 — APPROVED** [review]

**Pull request merged**